File: wpilibj/src/main/java/edu/wpi/first/wpilibj/simulation/AnalogTriggerSim.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package edu.wpi.first.wpilibj.simulation;

import edu.wpi.first.hal.simulation.AnalogTriggerDataJNI;
import edu.wpi.first.hal.simulation.NotifyCallback;
import edu.wpi.first.wpilibj.AnalogTrigger;
import java.util.NoSuchElementException;

/** Class to control a simulated analog trigger. */
public class AnalogTriggerSim {
  private final int m_index;

  /**
   * Constructs from an AnalogTrigger object.
   *
   * @param analogTrigger AnalogTrigger to simulate
   */
  public AnalogTriggerSim(AnalogTrigger analogTrigger) {
    m_index = analogTrigger.getIndex();
  }

  private AnalogTriggerSim(int index) {
    m_index = index;
  }

  /**
   * Creates an AnalogTriggerSim for an analog input channel.
   *
   * @param channel analog input channel
   * @return Simulated object
   * @throws NoSuchElementException if no AnalogTrigger is configured for that channel
   */
  public static AnalogTriggerSim createForChannel(int channel) {
    int index = AnalogTriggerDataJNI.findForChannel(channel);
    if (index < 0) {
      throw new NoSuchElementException("no analog trigger found for channel " + channel);
    }
    return new AnalogTriggerSim(index);
  }

  /**
   * Creates an AnalogTriggerSim for a simulated index. The index is incremented for each simulated
   * AnalogTrigger.
   *
   * @param index simulator index
   * @return Simulated object
   */
  public static AnalogTriggerSim createForIndex(int index) {
    return new AnalogTriggerSim(index);
  }

  /**
   * Register a callback on whether the analog trigger is initialized.
   *
   * @param callback the callback that will be called whenever the analog trigger is initialized
   * @param initialNotify if true, the callback will be run on the initial value
   * @return the {@link CallbackStore} object associated with this callback. Save a reference to
   *     this object so GC doesn't cancel the callback.
   */
  public CallbackStore registerInitializedCallback(NotifyCallback callback, boolean initialNotify) {
    int uid = AnalogTriggerDataJNI.registerInitializedCallback(m_index, callback, initialNotify);
    return new CallbackStore(m_index, uid, AnalogTriggerDataJNI::cancelInitializedCallback);
  }

  /**
   * Check if this analog trigger has been initialized.
   *
   * @return true if initialized
   */
  public boolean getInitialized() {
    return AnalogTriggerDataJNI.getInitialized(m_index);
  }

  /**
   * Change whether this analog trigger has been initialized.
   *
   * @param initialized the new value
   */
  public void setInitialized(boolean initialized) {
    AnalogTriggerDataJNI.setInitialized(m_index, initialized);
  }

  /**
   * Register a callback on the lower bound.
   *
   * @param callback the callback that will be called whenever the lower bound is changed
   * @param initialNotify if true, the callback will be run on the initial value
   * @return the {@link CallbackStore} object associated with this callback. Save a reference to
   *     this object so GC doesn't cancel the callback.
   */
  public CallbackStore registerTriggerLowerBoundCallback(
      NotifyCallback callback, boolean initialNotify) {
    int uid =
        AnalogTriggerDataJNI.registerTriggerLowerBoundCallback(m_index, callback, initialNotify);
    return new CallbackStore(m_index, uid, AnalogTriggerDataJNI::cancelTriggerLowerBoundCallback);
  }

  /**
   * Get the lower bound.
   *
   * @return the lower bound
   */
  public double getTriggerLowerBound() {
    return AnalogTriggerDataJNI.getTriggerLowerBound(m_index);
  }

  /**
   * Change the lower bound.
   *
   * @param triggerLowerBound the new lower bound
   */
  public void setTriggerLowerBound(double triggerLowerBound) {
    AnalogTriggerDataJNI.setTriggerLowerBound(m_index, triggerLowerBound);
  }

  /**
   * Register a callback on the upper bound.
   *
   * @param callback the callback that will be called whenever the upper bound is changed
   * @param initialNotify if true, the callback will be run on the initial value
   * @return the {@link CallbackStore} object associated with this callback. Save a reference to
   *     this object so GC doesn't cancel the callback.
   */
  public CallbackStore registerTriggerUpperBoundCallback(
      NotifyCallback callback, boolean initialNotify) {
    int uid =
        AnalogTriggerDataJNI.registerTriggerUpperBoundCallback(m_index, callback, initialNotify);
    return new CallbackStore(m_index, uid, AnalogTriggerDataJNI::cancelTriggerUpperBoundCallback);
  }

  /**
   * Get the upper bound.
   *
   * @return the upper bound
   */
  public double getTriggerUpperBound() {
    return AnalogTriggerDataJNI.getTriggerUpperBound(m_index);
  }

  /**
   * Change the upper bound.
   *
   * @param triggerUpperBound the new upper bound
   */
  public void setTriggerUpperBound(double triggerUpperBound) {
    AnalogTriggerDataJNI.setTriggerUpperBound(m_index, triggerUpperBound);
  }

  /** Reset all simulation data for this object. */
  public void resetData() {
    AnalogTriggerDataJNI.resetData(m_index);
  }
}

File: wpilibc/src/main/native/cpp/simulation/AnalogTriggerSim.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "frc/simulation/AnalogTriggerSim.h"

#include <memory>
#include <stdexcept>

#include <hal/simulation/AnalogTriggerData.h>

#include "frc/AnalogTrigger.h"

using namespace frc;
using namespace frc::sim;

AnalogTriggerSim::AnalogTriggerSim(const AnalogTrigger& analogTrigger)
    : m_index{analogTrigger.GetIndex()} {}

AnalogTriggerSim AnalogTriggerSim::CreateForChannel(int channel) {
  int index = HALSIM_FindAnalogTriggerForChannel(channel);
  if (index < 0) {
    throw std::out_of_range("no analog trigger found for channel");
  }
  return AnalogTriggerSim{index};
}

AnalogTriggerSim AnalogTriggerSim::CreateForIndex(int index) {
  return AnalogTriggerSim{index};
}

std::unique_ptr<CallbackStore> AnalogTriggerSim::RegisterInitializedCallback(
    NotifyCallback callback, bool initialNotify) {
  auto store = std::make_unique<CallbackStore>(
      m_index, -1, callback, &HALSIM_CancelAnalogTriggerInitializedCallback);
  store->SetUid(HALSIM_RegisterAnalogTriggerInitializedCallback(
      m_index, &CallbackStoreThunk, store.get(), initialNotify));
  return store;
}

bool AnalogTriggerSim::GetInitialized() const {
  return HALSIM_GetAnalogTriggerInitialized(m_index);
}

void AnalogTriggerSim::SetInitialized(bool initialized) {
  HALSIM_SetAnalogTriggerInitialized(m_index, initialized);
}

std::unique_ptr<CallbackStore>
AnalogTriggerSim::RegisterTriggerLowerBoundCallback(NotifyCallback callback,
                                                    bool initialNotify) {
  auto store = std::make_unique<CallbackStore>(
      m_index, -1, callback,
      &HALSIM_CancelAnalogTriggerTriggerLowerBoundCallback);
  store->SetUid(HALSIM_RegisterAnalogTriggerTriggerLowerBoundCallback(
      m_index, &CallbackStoreThunk, store.get(), initialNotify));
  return store;
}

double AnalogTriggerSim::GetTriggerLowerBound() const {
  return HALSIM_GetAnalogTriggerTriggerLowerBound(m_index);
}

void AnalogTriggerSim::SetTriggerLowerBound(double triggerLowerBound) {
  HALSIM_SetAnalogTriggerTriggerLowerBound(m_index, triggerLowerBound);
}

std::unique_ptr<CallbackStore>
AnalogTriggerSim::RegisterTriggerUpperBoundCallback(NotifyCallback callback,
                                                    bool initialNotify) {
  auto store = std::make_unique<CallbackStore>(
      m_index, -1, callback,
      &HALSIM_CancelAnalogTriggerTriggerUpperBoundCallback);
  store->SetUid(HALSIM_RegisterAnalogTriggerTriggerUpperBoundCallback(
      m_index, &CallbackStoreThunk, store.get(), initialNotify));
  return store;
}

double AnalogTriggerSim::GetTriggerUpperBound() const {
  return HALSIM_GetAnalogTriggerTriggerUpperBound(m_index);
}

void AnalogTriggerSim::SetTriggerUpperBound(double triggerUpperBound) {
  HALSIM_SetAnalogTriggerTriggerUpperBound(m_index, triggerUpperBound);
}

void AnalogTriggerSim::ResetData() {
  HALSIM_ResetAnalogTriggerData(m_index);
}
